### PR TITLE
[action] add NewEthSigner() and pass Signer as input to rlpRawHash() and rlpSignedHash()

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
Small pre-requisite code change to make (#3965) easier.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/web3server_utils.go [hunk 1]
```diff
@@ -117,7 +118,19 @@ func (svr *web3Handler) getTransactionFromActionInfo(blkHash hash.Hash256, selp
...
+		signer   types.Signer
+		encoding = iotextypes.Encoding(selp.Encoding())
+	)
+	if encoding != iotextypes.Encoding_IOTEX_PROTOBUF {
+		signer, err = action.NewEthSigner(encoding, svr.coreService.EVMNetworkID())
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		signer = types.HomesteadSigner{}
```
Comment (envestcc) [inline api/web3server_utils.go:131]: why use HomesteadSigner?

File: action/sealedenvelope.go [hunk 2]
```diff
@@ -134,8 +142,8 @@ func (sealed *SealedEnvelope) loadProto(pbAct *iotextypes.Action, evmID uint32)
 	if err != nil {
 		return err
 	}
-	encoding := pbAct.GetEncoding()
-	switch encoding {
+	sealed.encoding = pbAct.GetEncoding()
```
Comment (CoderZhi) [inline action/sealedenvelope.go:145]: even through errors are returned in line 151, 159, and 162, the encoding value of `sealed` has been modified. Unless we are not sure that the `sealed.encoding` won't be used any where else or the `sealed` instance won't be reused again.
Thus, we may want to introduce a local variable `encoding` here.

File: api/web3server_marshal_test.go [hunk 3]
```diff
@@ -155,7 +156,8 @@ func TestBlockObjectMarshal(t *testing.T) {
 
 	t.Run("BlockWithDetail", func(t *testing.T) {
 		raw := types.NewContractCreation(2, unit.ConvertIotxToRau(1000), 21000, unit.ConvertIotxToRau(1), []byte{})
-		ethTx, err := action.RawTxToSignedTx(raw, 0, sevlp.Signature())
+		signer, _ := action.NewEthSigner(iotextypes.Encoding_ETHEREUM_RLP, 0)
```
Comment (CoderZhi) [inline api/web3server_marshal_test.go:159]: `require.Nil(err)`

File: api/web3server_utils.go (hunk 1 above)
Comment (CoderZhi) [inline api/web3server_utils.go:131]: when will this happen?

File: action/rlp_tx.go [hunk 4]
```diff
@@ -9,17 +9,18 @@ import (
...
 	"github.com/iotexproject/go-pkgs/crypto"
 	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
 	"github.com/pkg/errors"
 	"golang.org/x/crypto/sha3"
 )
 
-func rlpRawHash(rawTx *types.Transaction, chainID uint32) (hash.Hash256, error) {
-	h := types.NewEIP155Signer(big.NewInt(int64(chainID))).Hash(rawTx)
+func rlpRawHash(rawTx *types.Transaction, signer types.Signer) (hash.Hash256, error) {
```
Comment (Liuhaai) [inline action/rlp_tx.go:17]: signer and hasher are of different roles

File: api/web3server_utils.go [hunk 5]
```diff
@@ -117,7 +118,19 @@ func (svr *web3Handler) getTransactionFromActionInfo(blkHash hash.Hash256, selp
...
+	}
```
Comment (Liuhaai) [inline api/web3server_utils.go:132]: revert this change in this pr and keep only refactoring part?

Comment (dustinxie) [reply api/web3server_utils.go:132]: make sense, this should be part of the next (#3965)

File: api/web3server_utils.go (hunk 1 above)
Comment (dustinxie) [reply api/web3server_utils.go:131]: this is for native tx (the signature format is same as Homestead)
I reverted as Haixiang mentioned, this part should be next PR

Comment (dustinxie) [reply api/web3server_utils.go:131]: this is for native tx (the signature format is same as Homestead)

File: action/rlp_tx.go (hunk 4 above)
Comment (dustinxie) [reply action/rlp_tx.go:17]: chatted offline, it's probably not worth the effort to define an interface `Hasher` for this, the code calling `signer.Hash()` is quite clear

File: action/sealedenvelope.go (hunk 2 above)
Comment (dustinxie) [reply action/sealedenvelope.go:145]: reverted

**Review by Liuhaai — APPROVED** [review]

File: action/rlp_tx.go [hunk 6]
```diff
@@ -88,5 +88,13 @@ func DecodeRawTx(rawData string, chainID uint32) (tx *types.Transaction, sig []b
 // NewEthSigner returns the proper signer for Eth-compatible tx
 func NewEthSigner(txType iotextypes.Encoding, chainID uint32) (types.Signer, error) {
```
Comment (CoderZhi) [inline action/rlp_tx.go:89]: if the signer of iotex format could be returned, the name of this function should be updated to `NewSigner`

**Review by CoderZhi — APPROVED** [review]
